lib/src/git.rs

```diff
@@ -213,28 +213,54 @@ pub enum GitExportError {
     InternalGitError(#[from] git2::Error),
 }
 
-/// Reflect changes made in the Jujutsu repo compared to our current view of the
-/// Git repo in `mut_repo.view().git_refs()`. Returns a list of names of
-/// branches that failed to export.
-// TODO: Also indicate why we failed to export these branches
+/// Export changes to branches made in the Jujutsu repo compared to the Git
+/// repo. Returns a list of names of branches that failed to export.
+///
+/// We ignore branches that are conflicted (were changed in the Git repo
+/// compared to our last remembered view of the Git repo stored on Git). These
+/// will be marked conflicted by the next `jj git import`.
+///
+/// We do not export tags and other refs at the moment, since these aren't
+/// supposed to be modified by JJ. For them, the Git state is considered
+/// authoritative.
+//
+// TODO: Also indicate the reason for any branches we failed to export
 pub fn export_refs(
     mut_repo: &mut MutableRepo,
     git_repo: &git2::Repository,
 ) -> Result<Vec<String>, GitExportError> {
-    // First find the changes we want need to make without modifying mut_repo
+    // It is important for `last_seen_view` to be unaffected by `jj undo`.
+    // Short-term TODO: this comment is expanded upon in a follow-up commit
+    let last_export_path = mut_repo.base_repo().repo_path().join("git_last_seen_refs");
+    let last_seen_view = crate::simple_op_store::read_view_from_file(last_export_path.clone())
+        .unwrap_or(crate::op_store::View::default());
+    let (new_exported_view, failed_branches) =
+        export_branches_unchanged_since_old_view(mut_repo, &View::new(last_seen_view), git_repo)?;
+    crate::simple_op_store::write_view_to_file(new_exported_view.store_view(), last_export_path)
+        .map_err(|err| GitExportError::StateReadWriteError(err.to_string()))?;
+    Ok(failed_branches)
+}
+
+/// Exports unconflicted branches.
+///
+/// Potential conflicts between the Git repo state and the JJ state are detected
+/// by using `old_view` as the conflict base. Returns the list of unexported
+/// branches and the new git state (with successfully exported branches moved).
+fn export_branches_unchanged_since_old_view(
+    mut_repo: &mut MutableRepo,
+    old_view: &View,
+    git_repo: &git2::Repository,
+) -> Result<(View, Vec<String>), GitExportError> {
+    let current_view = mut_repo.view();
+    let old_branches: HashSet<_> = old_view.branches().keys().cloned().collect();
+    let current_branches: HashSet<_> = current_view.branches().keys().cloned().collect(); // TODO: Do we need this?
```

Comment:
What do you mean by the TODO? The reason for the unioning below is so we detect added, modified, and removed branches (with `current_branches` needed for the "added" case, I think).

Reply:
I think I was wondering why to have these separately. We seem to only use the union. I'll probably add a commit combining them into one, and if it doesn't compile -- I'll know my answer.